ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -1825,25 +1825,38 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
         // Get entries starting from the token Key in this index.
         startTimeInMs = time.milliseconds();
         StoreFindToken newToken;
-        //if index based token points to the last entry of last index segment, and journal size cleaned up after last log
-        //segment auto close, we should return the original index based token instead of getting from journal.
-        if (isJournalEmptyAfterAutoClose() && indexBasedTokenPointsToLastEntryOfLastIndexSegment(storeToken,
-            indexSegments)) {
-          newToken = storeToken;
-        } else {
-          newToken = findEntriesFromSegmentStartOffset(storeToken.getOffset(), storeToken.getStoreKey(), messageEntries,
-              new FindEntriesCondition(maxTotalSizeOfEntries), indexSegments, storeToken.getInclusive());
-        }
-        logger.trace("Segment based token, Time used to find entries: {}", (time.milliseconds() - startTimeInMs));
+        long sizeSum = 0;
+        while (true) {
+          //if index based token points to the last entry of last index segment, and journal size cleaned up after last log
+          //segment auto close, we should return the original index based token instead of getting from journal.
+          if (isJournalEmptyAfterAutoClose() && indexBasedTokenPointsToLastEntryOfLastIndexSegment(storeToken,
+              indexSegments)) {
+            newToken = storeToken;
+          } else {
+            newToken =
+                findEntriesFromSegmentStartOffset(storeToken.getOffset(), storeToken.getStoreKey(), messageEntries,
+                    new FindEntriesCondition(maxTotalSizeOfEntries - sizeSum), indexSegments,
+                    storeToken.getInclusive());
+          }
+          logger.trace("Segment based token, Time used to find entries: {}", (time.milliseconds() - startTimeInMs));
 
-        startTimeInMs = time.milliseconds();
-        updateStateForMessages(messageEntries);
-        logger.trace("Segment based token, Time used to update state: {}", (time.milliseconds() - startTimeInMs));
+          startTimeInMs = time.milliseconds();
+          updateStateForMessages(messageEntries);
+          logger.trace("Segment based token, Time used to update state: {}", (time.milliseconds() - startTimeInMs));
 
-        startTimeInMs = time.milliseconds();
-        eliminateDuplicates(messageEntries);
-        logger.trace("Segment based token, Time used to eliminate duplicates: {}",
-            (time.milliseconds() - startTimeInMs));
+          startTimeInMs = time.milliseconds();
+          eliminateDuplicates(messageEntries);
+          logger.trace("Segment based token, Time used to eliminate duplicates: {}",
+              (time.milliseconds() - startTimeInMs));
+          sizeSum = messageEntries.stream().mapToLong(MessageInfo::getSize).sum();
```

Comment:
I don't see the code to determine the Blob is deleted or not. Or is that the MessageInfo::getSize different if the PutBlob is deleted?

Reply:
In findEntriesSince, there is a method call `updateStateForMessages`, it will update all the state for the put if the put is already deleted.